optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/ConstraintSession.java

```diff
@@ -40,15 +40,17 @@
      * As defined by {@link InnerScoreDirector#getConstraintMatchTotalMap()}.
      *
      * @return never null
+     * @param <Score_> the {@link Score} type
      */
-    Map<String, ConstraintMatchTotal> getConstraintMatchTotalMap();
+    <Score_ extends Score<Score_>> Map<String, ConstraintMatchTotal<Score_>> getConstraintMatchTotalMap();
```

Comment:
Considering how often these two methods (constraint match total map, indictment map) repeat in various unrelated interfaces, perhaps these should be extracted into their own interface, which the other interfaces would then extend?

Reply:
I think this is premature optimizaton, that if we clean up some architecture, will become much less.

They repeat 3 times:
- ScoreExplenation - which is public API, so the extra interface inheritence will complicate it for users
- ScoreDirector
- ConstraintSession

I believe ConstraintSession is a needless indirection (that maybe even has a perf impact).
Same for ConstraintSessionFactory and ConstraintStreamScoreDirectorFactory.
We could maybe replace it by [Drools/Bavet]ConstraintStreamScoreDirector(Factory) extends AbstractConstraintStreamScoreDirector(Factory).
That experiment needs to be looked into first.